src/mlpack/core/tree/binary_space_tree/rp_tree_max_split_impl.hpp

```diff
@@ -0,0 +1,156 @@
+bool RPTreeMaxSplit<BoundType, MatType>::SplitNode(const BoundType& /* bound */,
+                                                  MatType& data,
+                                                  const size_t begin,
+                                                  const size_t count,
+                                                  SplitInfo& splitInfo)
+{
+  splitInfo.direction.zeros(data.n_rows);
+
+  // Get the normal to the hyperplane.
+  GetRandomDirection(splitInfo.direction);
+
+  // Get the value according to which we will perform the split.
+  if (!GetSplitVal(data, begin, count, splitInfo.direction, splitInfo.splitVal))
+    return false;
+
+  return true;
+}
+
+template<typename BoundType, typename MatType>
+void RPTreeMaxSplit<BoundType, MatType>::GetRandomDirection(
+    arma::Col<ElemType>& direction)
+{
+  arma::Col<ElemType> origin;
+
+  origin.zeros(direction.n_rows);
+
+  for (size_t k = 0; k < direction.n_rows; k++)
+    direction[k] = math::Random(-1.0, 1.0);
+
+  ElemType length = metric::EuclideanDistance::Evaluate(origin, direction);
+
+  if (length > 0)
+    direction /= length;
+  else
+  {
+    // If the vector is equal to 0, choose an arbitrary dimension.
+    size_t k = math::RandInt(direction.n_rows);
+
+    direction[k] = 1.0;
+
+    length = metric::EuclideanDistance::Evaluate(origin, direction);
+
+    direction[k] /= length;
+  }
+}
+
+template<typename BoundType, typename MatType>
+typename MatType::elem_type RPTreeMaxSplit<BoundType, MatType>::
+GetRandomDeviation(const MatType& data,
+                   const size_t begin,
+                   const size_t count,
+                   const arma::Col<ElemType>& direction)
+{
+  // Choose a random point
+  size_t index = math::RandInt(begin, begin + count);
+
+  ElemType furthestDistance = 0;
+
+  // Find the furthest point from the point that we chose
+  for (size_t i = begin; i < index; i++)
+  {
+    const ElemType dist = metric::SquaredEuclideanDistance::Evaluate(
+        data.col(index), data.col(i));
+    if (dist > furthestDistance)
+      furthestDistance = dist;
+  }
+
+  for (size_t i = index; i < begin + count; i++)
+  {
+    const ElemType dist = metric::SquaredEuclideanDistance::Evaluate(
+        data.col(index), data.col(i));
+    if (dist > furthestDistance)
+      furthestDistance = dist;
+  }
+
+  // Get a random deviation.
+  return math::Random(-6.0 * std::sqrt(furthestDistance / data.n_rows),
+                      6.0 * std::sqrt(furthestDistance / data.n_rows));
+}
+
+template<typename BoundType, typename MatType>
+void RPTreeMaxSplit<BoundType, MatType>::GetDistinctSamples(
+    arma::uvec& distinctSamples,
+    const size_t begin,
+    const size_t count,
+    const size_t numSamples)
+{
+  arma::Col<size_t> samples;
+
+  samples.zeros(count);
+
+  for (size_t i = 0; i < numSamples; i++)
+    samples [ (size_t) math::RandInt(count) ]++;
+
+  distinctSamples = arma::find(samples > 0);
+
+  distinctSamples += begin;
+}
```

Comment:
Yes, I remember. That branch has not merged yet. I think it is better to correct that after merging.